### PR TITLE
Hgi 393/setup optimizely action destination

### PR DESCRIPTION
**Pull request by hvardhan-unth**
This pull request includes the setup of Optimizely cloud new `action-destinations `  without implementation of LRU-cache.
It also includes unit test cases.

JIRA ticket link:- https://segment.atlassian.net/browse/HGI-393

## Testing

Testing completed successfully on local.

Screenshot of Action tester

Screenshot 2023-04-05 at 12 13 39 PM

Screenshot 2023-04-05 at 12 14 17 PM

### CODE REVIEW

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/index.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      }
+    },
+    testAuthentication: (request, { settings }) => {
+      const { dataFileUrl } = settings
+      return request(dataFileUrl)
+    }
+  },
+  onDelete: async (request, { settings }) => {
+    if (!settings.accessToken) {
+      throw new IntegrationError('Access Token is required for user deletion.')
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/index.ts:44]: Can we provide status for IntegrationError as 400?

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/dataFile.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
+export const dataFile = {
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/dataFile.ts:1]: Can you move this file to __tests__ folder?

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/index.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+      }
+    }
+  },
+  perform: async (request, { settings, payload }) => {
+    const result = await request<ProjectConfig>(settings.dataFileUrl)
+    const dataFile = result.data
+    const eventId = getEventId(dataFile, payload.eventKey)
+
+    if (!eventId) {
+      throw new IntegrationError(`Event with name ${payload.eventKey} is not defined`)
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/index.ts:111]:
```suggestion
      throw new PayloadValidationError(`Event with name ${payload.eventKey} is not defined`)
```

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/index.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+      })
+    }
+
+    await request('https://logx.optimizely.com/v1/events', {
+      method: 'POST',
+      json: {
+        account_id: dataFile.accountId,
+        anonymize_ip: dataFile.anonymizeIP,
+        client_name: 'Segment',
+        client_version: '1.0.01',
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/index.ts:153]: Is client_version mandatory?

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
+import nock from 'nock'
+import { createTestEvent, createTestIntegration } from '@segment/actions-core'
+import Destination from '../../index'
+import { dataFile } from '../dataFile'
+
+const testDestination = createTestIntegration(Destination)
+
+describe('OptimizelyFullStack.trackEvent', () => {
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts:8]: Can we cover all the control flows? Right now, it seems like only 2 are covered.

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/index.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  name: 'Optimizely Feature Experimentation (Actions)',
+  slug: 'actions-optimizely-feature-experimentation-cloud',
+  mode: 'cloud',
+
+  authentication: {
+    scheme: 'custom',
+    fields: {
+      accessToken: {
+        label: 'Personal Access Token',
+        description: 'Personal Access Token which client generated manually.',
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/index.ts:17]: @varadarajan-tw - to review and provide appropriate descriptions.

**Review by varadarajan-tw — COMMENTED** [review]
Great work! Left a few comments. Lets ensure that all the fields we pass appear appropriately in Optimizely!

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/index.ts (hunk 1 above)
Comment (hvardhan-unth) [reply packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/index.ts]: Yes

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/dataFile.ts (hunk 2 above)
Comment (hvardhan-unth) [reply packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/dataFile.ts:1]: If I am moving this file in test folder, it is giving error that file must contain a single test case in test suite.

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/index.ts (hunk 4 above)
Comment (hvardhan-unth) [reply packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/index.ts]: It is not mandatory, so removed this.

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/dataFile.ts (hunk 2 above)
Comment (varadarajan-tw) [reply packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/dataFile.ts:1]: Okay. Seems like Jest by default considers anything in __tests__ folder as tests. Can we atleast name it is mock-datafile.ts?

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        traits: {
+          test: 'test'
+        }
+      }
+    })
+    await expect(
+      testDestination.testAction('trackEvent', { event, settings, useDefaultMappings: true })
+    ).resolves.not.toThrowError()
+  }),
+    it('should throw error if event sent is not in datafile', async () => {
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts:31]: I don't see any new tests added? Is the commit missing?

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/index.ts (hunk 6 above)
Comment (varadarajan-tw) [reply packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/index.ts:17]: Descriptions:
`Personal Access Token` - The personal access token will be used to submit a [GDPR delete](https://segment.com/docs/privacy/user-deletion-and-suppression/) request to Optimizely. To generate a personal access token, navigate to [Profile->API Access](https://app.optimizely.com/v2/profile/api) and generate a new token.

`Datafile URL` - The datafile is a JSON representation of the current state of flags and experiments for an environment in your Full Stack project. It should look something like https://cdn.optimizely.com/json/9218021209.json

**Review by varadarajan-tw — COMMENTED** [review]
Left few more comments.

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/index.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        type: 'password',
+        required: false
+      },
+      accountId: {
```
[the remark this answers is not in the document]
Comment (hvardhan-unth) [reply packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/index.ts:21]: GDPR delete functionality also requires the accountId, which is getting from settings.

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/index.ts [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'Track Event',
+  description: 'Track Conversion Events',
+  defaultSubscription: 'type = "track"',
+  fields: {
+    eventKey: {
+      label: 'Event Key',
+      type: 'string',
+      description:
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/index.ts:16]: As per [SDD](https://docs.google.com/document/d/1FKjNSSNouHaVf7BjS3OacJOCfQhhrEseW1GWRsjCqH4/edit#
), this should be a dynamic field. Can we address this either in this PR or next one? Pls take a look at Segment Profiles.

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/index.ts [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+        'The ID of the user associated with the event being tracked. **Important**: This ID must match the user ID provided to Activate or Is Feature Enabled.',
+      required: true,
+      default: {
+        '@if': {
+          exists: {
+            '@path': '$.userId'
+          },
+
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/index.ts:34]:
```suggestion
```

**Review by varadarajan-tw — COMMENTED** [review]
Thank you for doing the recommended changes. Overall it looks in good shape. If you can address comments to the tests, i'll be happy to give an approval for merging this.

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/index.ts [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+  description: 'Track Conversion Events',
+  defaultSubscription: 'type = "track"',
+  fields: {
+    eventKey: {
+      label: 'Event Key',
+      type: 'string',
+      description:
+        'The key of the event to be tracked. This key must match the event key provided when the event was created in the Optimizely app.',
+      required: true,
+      dynamic: true,
```
[the remark this answers is not in the document]
Comment (hvardhan-unth) [reply packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/index.ts:19]: done

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        }
+      }
+    })
+    await expect(
+      testDestination.testAction('trackEvent', { event, settings, useDefaultMappings: true })
+    ).resolves.not.toThrowError()
+  })
+})
+
+describe('.getEventId', () => {
```
[the remark this answers is not in the document]
Comment (hvardhan-unth) [reply packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts]: sure

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  it('should send event successfully', async () => {
+    const settings = {
+      accountId: '12345566',
+      dataFileUrl: 'https://cdn.example.com/dataFile.json'
+    }
+    nock(settings.dataFileUrl).get('').reply(200, dataFile)
+    nock('https://logx.optimizely.com/v1/events').post('').reply(200)
+
+    const res = await got.post('https://logx.optimizely.com/v1/events', { json: payload })
+    expect(res.statusCode).toBe(200)
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts:20]: @hvardhan-unth - I still don't see payload assertions added! 😅

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+      properties: {
+        revenue: 1000
+      },
+      context: {
+        traits: {
+          test: 'test'
+        }
+      }
+    })
+    await expect(
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts:63]: How do we validate bot filtering worked in this test? Can you add assertions on the payload sent to the destination?

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+describe('OptimizelyFeatureExperimentation.trackEvent', () => {
+  it('should send event successfully', async () => {
+    const settings = {
+      accountId: '12345566',
+      dataFileUrl: 'https://cdn.example.com/dataFile.json'
+    }
+    nock(settings.dataFileUrl).get('').reply(200, dataFile)
+    nock('https://logx.optimizely.com/v1/events').post('').reply(200)
+
+    const res = await got.post('https://logx.optimizely.com/v1/events', { json: payload })
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts:19]: why are posting to `got` here? We should be calling testAction right?

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts (hunk 13 above)
Comment (varadarajan-tw) [reply packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts:20]: Example - https://github.com/segmentio/action-destinations/blob/78fbe5de55c5b104a76bbdc74f27192368ba8a69/packages/destination-actions/src/destinations/google-enhanced-conversions/__tests__/uploadCallConversion.test.ts#L92

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/index.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (hvardhan-unth) [reply packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/index.ts:19]: It is written in function.test.ts file:- [https://github.com/segmentio/action-destinations/pull/1153/files#diff-0b902234976a04e31751d20ee7d046136fd1d2578c1e52b2c2edffb81f0ab068R21](url)

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts (hunk 13 above)
Comment (hvardhan-unth) [reply packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts:20]: Now, done 😅

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts (hunk 15 above)
Comment (hvardhan-unth) [reply packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts]: Removed.

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts (hunk 14 above)
Comment (hvardhan-unth) [reply packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/index.test.ts]: Done

File: packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/functions.test.ts [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+describe('.getEventId', () => {
+  it('should return eventId for eventKey', async () => {
+    expect(getEventId(dataFile, 'Product List Clicked')).toBe('22020998834')
+  })
+})
+
+describe('.buildVisitorAttributes', () => {
+  it('should return visitor attributes for payload', async () => {
+    expect(buildVisitorAttributes(dataFile, { id: '18531090301', key: 'test' })).toStrictEqual([])
```
[the remark this answers is not in the document]
Comment (hvardhan-unth) [reply packages/destination-actions/src/destinations/optimizely-feature-experimentation-actions/trackEvent/__tests__/functions.test.ts]: Updated the test case.

**Review by varadarajan-tw — APPROVED** [review]
Thanks for implementing the suggested. This looks good.
Confirmed offline that LRU cache change is to be updated in Separate PR.

**Pull request merged**